runners/flink/src/test/java/org/apache/beam/runners/flink/PortableExecutionTest.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.core.construction.PipelineTranslation;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.ServerFactory;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService;
+import org.apache.beam.runners.fnexecution.control.ControlClientPool.Source;
+import org.apache.beam.runners.fnexecution.control.DockerJobBundleFactory;
+import org.apache.beam.runners.fnexecution.control.FnApiControlClientPoolService;
+import org.apache.beam.runners.fnexecution.control.InstructionRequestHandler;
+import org.apache.beam.runners.fnexecution.environment.EnvironmentFactory;
+import org.apache.beam.runners.fnexecution.environment.RemoteEnvironment;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
+import org.apache.beam.runners.fnexecution.provisioning.StaticGrpcProvisionService;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.BigEndianLongCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.fn.IdGenerator;
+import org.apache.beam.sdk.fn.channel.ManagedChannelFactory;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.Impulse;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.values.KV;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Tests the execution of a pipeline from specification to execution on the portable Flink runner.
+ * Exercises job invocation, executable stage translation and deployment with embedded Flink for
+ * batch and streaming.
+ */
+@RunWith(Parameterized.class)
+public class PortableExecutionTest implements Serializable {
+
+  @Parameters
+  public static Object[] data() {
+    return new Object[] {true, false};
+  }
+
+  @Parameter public boolean isStreaming;
+
+  private transient ExecutorService sdkHarnessExecutor;
+  private transient Future<?> sdkHarnessExecutorFuture;
+
+  private DockerJobBundleFactory createJobBundleFactory(JobInfo jobInfo) throws Exception {
+    return new DockerJobBundleFactory(jobInfo) {
+
+      @Override
+      protected ServerFactory getServerFactory() {
+        return ServerFactory.createWithUrlFactory(
+            (host, port) -> HostAndPort.fromParts("localhost", port).toString());
+        //return InProcessServerFactory.create();
+      }
+
+      @Override
+      protected EnvironmentFactory getEnvironmentFactory(
```

Comment:
This is really useful for in process execution instead of over docker.
Shall we put it in main code?

Reply:
That's a good idea. If there is more interest in this we can do that as follow-up. @bsidhom what do you think?

Reply:
Please re-use the existing `InProcessEnvironmentFactory` since it does a good portion of what you want to do here.

Also consider taking a look at `InProcessSdkHarness` which is a test rule which spins up an SDK harness in process. I don't think it will be overly useful though unless you can get a good way to inject the `SdkHarnessClient` that it provides into the Flink runtime instead of the current method of setting a specific instance in the `DockerJobBundleFactory`.

Reply:
Super, `InProcessEnvironmentFactory` works like a charm and cuts execution time to half! `InProcessSdkHarness` isn't useful in this case since the services are provided by the Flink runner and the environment needs to be injected.